File: japiwzsk/PnlWzskNavOp.java
/**
  * \file PnlWzskNavOp.java
  * Java API code for job PnlWzskNavOp
  * \author Catherine Johnson
  * \date created: 6 Oct 2020
  * \date modified: 6 Oct 2020
  */

package apiwzsk;

import java.util.*;
import org.w3c.dom.*;
import sbecore.*;

public class PnlWzskNavOp {
	/**
		* VecVDo (full: VecVWzskNavOpDo)
		*/
	public static class VecVDo {

		public static final int BUTLLVNEWCRDCLICK = 1;
		public static final int BUTLIVNEWCRDCLICK = 2;

		public static int getIx(
					String sref
				) {
			String s = sref.toLowerCase();

			if (s.equals("butllvnewcrdclick")) return BUTLLVNEWCRDCLICK;
			if (s.equals("butlivnewcrdclick")) return BUTLIVNEWCRDCLICK;

			return 0;
		};

		public static String getSref(
					int ix
				) {
			if (ix == BUTLLVNEWCRDCLICK) return("ButLlvNewcrdClick");
			if (ix == BUTLIVNEWCRDCLICK) return("ButLivNewcrdClick");

			return "";
		};

	};

	/**
	  * StatApp (full: StatAppWzskNavOp)
	  */
	public class StatApp extends Block {

		public static final int IXWZSKVEXPSTATE = 1;

		public StatApp(
					int ixWzskVExpstate
				) {
			this.ixWzskVExpstate = ixWzskVExpstate;

			mask = new HashSet<Integer>(Arrays.asList(IXWZSKVEXPSTATE));
		};

		public int ixWzskVExpstate;

		public boolean readXML(
					Document doc
					, String basexpath
					, boolean addbasetag
				) {
			String srefIxWzskVExpstate;

			clear();

			if (addbasetag) basexpath = Xmlio.checkUclcXPaths(doc, basexpath, "StatAppWzskNavOp");

			String itemtag = "StatitemAppWzskNavOp";

			if (Xmlio.checkXPath(doc, basexpath)) {
				srefIxWzskVExpstate = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Si", "sref", "srefIxWzskVExpstate", mask, IXWZSKVEXPSTATE);
				ixWzskVExpstate = VecWzskVExpstate.getIx(srefIxWzskVExpstate);

				return true;
			};

			return false;
		};

		public HashSet<Integer> comm(
					StatApp comp
				) {
			HashSet<Integer> items = new HashSet<Integer>();

			if (ixWzskVExpstate == comp.ixWzskVExpstate) items.add(IXWZSKVEXPSTATE);

			return(items);
		};

		public HashSet<Integer> diff(
					StatApp comp
				) {
			HashSet<Integer> commitems;
			HashSet<Integer> diffitems;

			commitems = comm(comp);

			diffitems = new HashSet<Integer>(Arrays.asList(IXWZSKVEXPSTATE));
			for (Integer ci: commitems) diffitems.remove(ci);

			return(diffitems);
		};

	};

	/**
	  * StatShr (full: StatShrWzskNavOp)
	  */
	public class StatShr extends Block {

		public static final int BUTLLVNEWCRDAVAIL = 1;
		public static final int BUTLIVNEWCRDAVAIL = 2;

		public StatShr(
					boolean ButLlvNewcrdAvail
					, boolean ButLivNewcrdAvail
				) {
			this.ButLlvNewcrdAvail = ButLlvNewcrdAvail;
			this.ButLivNewcrdAvail = ButLivNewcrdAvail;

			mask = new HashSet<Integer>(Arrays.asList(BUTLLVNEWCRDAVAIL, BUTLIVNEWCRDAVAIL));
		};

		public boolean ButLlvNewcrdAvail;
		public boolean ButLivNewcrdAvail;

		public boolean readXML(
					Document doc
					, String basexpath
					, boolean addbasetag
				) {

			clear();

			if (addbasetag) basexpath = Xmlio.checkUclcXPaths(doc, basexpath, "StatShrWzskNavOp");

			String itemtag = "StatitemShrWzskNavOp";

			if (Xmlio.checkXPath(doc, basexpath)) {
				ButLlvNewcrdAvail = Xmlio.extractBooleanAttrUclc(doc, basexpath, itemtag, "Si", "sref", "ButLlvNewcrdAvail", mask, BUTLLVNEWCRDAVAIL);
				ButLivNewcrdAvail = Xmlio.extractBooleanAttrUclc(doc, basexpath, itemtag, "Si", "sref", "ButLivNewcrdAvail", mask, BUTLIVNEWCRDAVAIL);

				return true;
			};

			return false;
		};

		public HashSet<Integer> comm(
					StatShr comp
				) {
			HashSet<Integer> items = new HashSet<Integer>();

			if (ButLlvNewcrdAvail == comp.ButLlvNewcrdAvail) items.add(BUTLLVNEWCRDAVAIL);
			if (ButLivNewcrdAvail == comp.ButLivNewcrdAvail) items.add(BUTLIVNEWCRDAVAIL);

			return(items);
		};

		public HashSet<Integer> diff(
					StatShr comp
				) {
			HashSet<Integer> commitems;
			HashSet<Integer> diffitems;

			commitems = comm(comp);

			diffitems = new HashSet<Integer>(Arrays.asList(BUTLLVNEWCRDAVAIL, BUTLIVNEWCRDAVAIL));
			for (Integer ci: commitems) diffitems.remove(ci);

			return(diffitems);
		};

	};

	/**
	  * Tag (full: TagWzskNavOp)
	  */
	public class Tag extends Block {

		public static final int CPT = 1;
		public static final int CPTLLV = 2;
		public static final int CPTLIV = 3;

		public Tag(
					String Cpt
					, String CptLlv
					, String CptLiv
				) {
			this.Cpt = Cpt;
			this.CptLlv = CptLlv;
			this.CptLiv = CptLiv;

			mask = new HashSet<Integer>(Arrays.asList(CPT, CPTLLV, CPTLIV));
		};

		public String Cpt;
		public String CptLlv;
		public String CptLiv;

		public boolean readXML(
					Document doc
					, String basexpath
					, boolean addbasetag
				) {

			clear();

			if (addbasetag) basexpath = Xmlio.checkUclcXPaths(doc, basexpath, "TagWzskNavOp");

			String itemtag = "TagitemWzskNavOp";

			if (Xmlio.checkXPath(doc, basexpath)) {
				Cpt = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Ti", "sref", "Cpt", mask, CPT);
				CptLlv = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Ti", "sref", "CptLlv", mask, CPTLLV);
				CptLiv = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Ti", "sref", "CptLiv", mask, CPTLIV);

				return true;
			};

			return false;
		};

		public HashSet<Integer> comm(
					Tag comp
				) {
			HashSet<Integer> items = new HashSet<Integer>();

			if (Cpt.equals(comp.Cpt)) items.add(CPT);
			if (CptLlv.equals(comp.CptLlv)) items.add(CPTLLV);
			if (CptLiv.equals(comp.CptLiv)) items.add(CPTLIV);

			return(items);
		};

		public HashSet<Integer> diff(
					Tag comp
				) {
			HashSet<Integer> commitems;
			HashSet<Integer> diffitems;

			commitems = comm(comp);

			diffitems = new HashSet<Integer>(Arrays.asList(CPT, CPTLLV, CPTLIV));
			for (Integer ci: commitems) diffitems.remove(ci);

			return(diffitems);
		};

	};

	/**
		* DpchAppDo (full: DpchAppWzskNavOpDo)
		*/
	public class DpchAppDo extends DpchAppWzsk {

		public static final int SCRJREF = 1;
		public static final int IXVDO = 2;
		public static final int ALL = 3;

		public DpchAppDo(
					String scrJref
					, int ixVDo
					, Integer[] mask
				) {
			super(VecWzskVDpch.DPCHAPPWZSKNAVOPDO, scrJref);

			this.mask = new HashSet<Integer>(Arrays.asList(mask));

			for (Integer i: mask)
				if (i == ALL) {
					this.mask = new HashSet<Integer>(Arrays.asList(SCRJREF, IXVDO));
					break;
				};

			this.ixVDo = ixVDo;
		};

		public int ixVDo;

		public String getSrefsMask() {
			ArrayList<String> ss = new ArrayList<String>();

			if (has(SCRJREF)) ss.add("scrJref");
			if (has(IXVDO)) ss.add("ixVDo");

			return StrMod.vectorToString(ss, ';');
		};

		public void writeXML(
					Document doc
					, Element sup
				) {
			Element el = doc.createElement("DpchAppWzskNavOpDo");

			if (sup == null) doc.appendChild(el);
			else sup.appendChild(el);

			el.setAttribute("xmlns", "http://www.mpsitech.com");

			if (has(SCRJREF)) Xmlio.writeString(doc, el, "scrJref", scrJref);
			if (has(IXVDO)) Xmlio.writeString(doc, el, "srefIxVDo", VecVDo.getSref(ixVDo));
		};

	};

	/**
		* DpchEngData (full: DpchEngWzskNavOpData)
		*/
	public class DpchEngData extends DpchEngWzsk {

		public static final int SCRJREF = 1;
		public static final int STATAPP = 2;
		public static final int STATSHR = 3;
		public static final int TAG = 4;

		public DpchEngData() {
			super(VecWzskVDpch.DPCHENGWZSKNAVOPDATA);

			statapp = new StatApp(0);
			statshr = new StatShr(false, false);
			tag = new Tag("", "", "");
		};

		public StatApp statapp;
		public StatShr statshr;
		public Tag tag;

		public String getSrefsMask() {
			ArrayList<String> ss = new ArrayList<String>();

			if (has(SCRJREF)) ss.add("scrJref");
			if (has(STATAPP)) ss.add("statapp");
			if (has(STATSHR)) ss.add("statshr");
			if (has(TAG)) ss.add("tag");

			return StrMod.vectorToString(ss, ';');
		};

		public void readXML(
					Document doc
					, String basexpath
					, boolean addbasetag
				) {

			clear();

			if (addbasetag) basexpath = Xmlio.checkUclcXPaths(doc, basexpath, "DpchEngWzskNavOpData");

			if (Xmlio.checkXPath(doc, basexpath)) {
				scrJref = Xmlio.extractStringUclc(doc, basexpath, "scrJref", "", mask, SCRJREF);
				if (statapp.readXML(doc, basexpath, true)) add(STATAPP);
				if (statshr.readXML(doc, basexpath, true)) add(STATSHR);
				if (tag.readXML(doc, basexpath, true)) add(TAG);
			} else {
				scrJref = "";
				statapp = new StatApp(0);
				statshr = new StatShr(false, false);
				tag = new Tag("", "", "");
			};
		};

	};

};

File: swapiwzsk/PnlWzskNavOp.swift
/**
  * \file PnlWzskNavOp.swift
  * Swift API code for job PnlWzskNavOp
	* \copyright (C) 2022 MPSI Technologies GmbH
	* \author Alexander Wirthmueller (auto-generation)
	* \date created: 12 Sep 2022
	*/
// IP header --- ABOVE

import Foundation
import sbecore

/**
	* PnlWzskNavOp
	*/
public class PnlWzskNavOp {
	/**
		* VecVDo (full: VecVWzskNavOpDo)
		*/
	public class VecVDo {

		public static let BUTLLVNEWCRDCLICK: Int = 1
		public static let BUTLIVNEWCRDCLICK: Int = 2

		public static func getIx(
					_ sref: String
				) -> Int {
			let s: String = sref.lowercased()

			if s == "butllvnewcrdclick" {return BUTLLVNEWCRDCLICK}
			if s == "butlivnewcrdclick" {return BUTLIVNEWCRDCLICK}

			return 0
		}

		public static func getSref(
					_ ix: Int
				) -> String {
			if ix == BUTLLVNEWCRDCLICK {return "ButLlvNewcrdClick"}
			if ix == BUTLIVNEWCRDCLICK {return "ButLivNewcrdClick"}

			return ""
		}

	}

	/**
	  * StatApp (full: StatAppWzskNavOp)
	  */
	public class StatApp: Block {

		public static let IXWZSKVEXPSTATE: Int = 1

		public init(
					_ ixWzskVExpstate: Int = 0
				) {
			self.ixWzskVExpstate = ixWzskVExpstate

			super.init()

			mask = [StatApp.IXWZSKVEXPSTATE]
		}

		public var ixWzskVExpstate: Int

		public func readXML(
					_ doc: XMLDocument
					, _ _basexpath: String
					, _ addbasetag: Bool
				) -> Bool {
			var valid: Bool
			var basexpath: String? = _basexpath

			var srefIxWzskVExpstate: String

			clear()

			if addbasetag {basexpath = Xmlio.checkUclcXPaths(doc, _basexpath, "StatAppWzskNavOp")}
			valid = basexpath != nil

			let itemtag: String = "StatitemAppWzskNavOp"

			if valid {
				srefIxWzskVExpstate = Xmlio.extractStringAttrUclc(doc, basexpath!, itemtag, "Ci", "sref", "srefIxWzskVExpstate", &mask, StatApp.IXWZSKVEXPSTATE)
				ixWzskVExpstate = VecWzskVExpstate.getIx(srefIxWzskVExpstate)
			}

			return valid
		}

		public func comm(
					_ comp: StatApp
				) -> Set<Int> {
			var items: Set<Int> = []

			if ixWzskVExpstate == comp.ixWzskVExpstate {items.insert(StatApp.IXWZSKVEXPSTATE)}

			return items
		}

		public func diff(
					_ comp: StatApp
				) -> Set<Int> {
			let commitems: Set<Int> = comm(comp)
			var diffitems: Set<Int> = [StatApp.IXWZSKVEXPSTATE]

			for ci in commitems {diffitems.remove(ci)}

			return diffitems
		}

	}

	/**
	  * StatShr (full: StatShrWzskNavOp)
	  */
	public class StatShr: Block {

		public static let BUTLLVNEWCRDAVAIL: Int = 1
		public static let BUTLIVNEWCRDAVAIL: Int = 2

		public init(
					_ ButLlvNewcrdAvail: Bool = false
					, _ ButLivNewcrdAvail: Bool = false
				) {
			self.ButLlvNewcrdAvail = ButLlvNewcrdAvail
			self.ButLivNewcrdAvail = ButLivNewcrdAvail

			super.init()

			mask = [StatShr.BUTLLVNEWCRDAVAIL, StatShr.BUTLIVNEWCRDAVAIL]
		}

		public var ButLlvNewcrdAvail: Bool
		public var ButLivNewcrdAvail: Bool

		public func readXML(
					_ doc: XMLDocument
					, _ _basexpath: String
					, _ addbasetag: Bool
				) -> Bool {
			var valid: Bool
			var basexpath: String? = _basexpath

			clear()

			if addbasetag {basexpath = Xmlio.checkUclcXPaths(doc, _basexpath, "StatShrWzskNavOp")}
			valid = basexpath != nil

			let itemtag: String = "StatitemShrWzskNavOp"

			if valid {
				ButLlvNewcrdAvail = Xmlio.extractBoolAttrUclc(doc, basexpath!, itemtag, "Ci", "sref", "ButLlvNewcrdAvail", &mask, StatShr.BUTLLVNEWCRDAVAIL)
				ButLivNewcrdAvail = Xmlio.extractBoolAttrUclc(doc, basexpath!, itemtag, "Ci", "sref", "ButLivNewcrdAvail", &mask, StatShr.BUTLIVNEWCRDAVAIL)
			}

			return valid
		}

		public func comm(
					_ comp: StatShr
				) -> Set<Int> {
			var items: Set<Int> = []

			if ButLlvNewcrdAvail == comp.ButLlvNewcrdAvail {items.insert(StatShr.BUTLLVNEWCRDAVAIL)}
			if ButLivNewcrdAvail == comp.ButLivNewcrdAvail {items.insert(StatShr.BUTLIVNEWCRDAVAIL)}

			return items
		}

		public func diff(
					_ comp: StatShr
				) -> Set<Int> {
			let commitems: Set<Int> = comm(comp)
			var diffitems: Set<Int> = [StatShr.BUTLLVNEWCRDAVAIL, StatShr.BUTLIVNEWCRDAVAIL]

			for ci in commitems {diffitems.remove(ci)}

			return diffitems
		}

	}

	/**
	  * Tag (full: TagWzskNavOp)
	  */
	public class Tag: Block {

		public static let CPT: Int = 1
		public static let CPTLLV: Int = 2
		public static let CPTLIV: Int = 3

		public init(
					_ Cpt: String = ""
					, _ CptLlv: String = ""
					, _ CptLiv: String = ""
				) {
			self.Cpt = Cpt
			self.CptLlv = CptLlv
			self.CptLiv = CptLiv

			super.init()

			mask = [Tag.CPT, Tag.CPTLLV, Tag.CPTLIV]
		}

		public var Cpt: String
		public var CptLlv: String
		public var CptLiv: String

		public func readXML(
					_ doc: XMLDocument
					, _ _basexpath: String
					, _ addbasetag: Bool
				) -> Bool {
			var valid: Bool
			var basexpath: String? = _basexpath

			clear()

			if addbasetag {basexpath = Xmlio.checkUclcXPaths(doc, _basexpath, "TagWzskNavOp")}
			valid = basexpath != nil

			let itemtag: String = "TagitemzskNavOp"

			if valid {
				Cpt = Xmlio.extractStringAttrUclc(doc, basexpath!, itemtag, "Ti", "sref", "Cpt", &mask, Tag.CPT)
				CptLlv = Xmlio.extractStringAttrUclc(doc, basexpath!, itemtag, "Ti", "sref", "CptLlv", &mask, Tag.CPTLLV)
				CptLiv = Xmlio.extractStringAttrUclc(doc, basexpath!, itemtag, "Ti", "sref", "CptLiv", &mask, Tag.CPTLIV)
			}

			return valid
		}

		public func comm(
					_ comp: Tag
				) -> Set<Int> {
			var items: Set<Int> = []

			if Cpt == comp.Cpt {items.insert(Tag.CPT)}
			if CptLlv == comp.CptLlv {items.insert(Tag.CPTLLV)}
			if CptLiv == comp.CptLiv {items.insert(Tag.CPTLIV)}

			return items
		}

		public func diff(
					_ comp: Tag
				) -> Set<Int> {
			let commitems: Set<Int> = comm(comp)
			var diffitems: Set<Int> = [Tag.CPT, Tag.CPTLLV, Tag.CPTLIV]

			for ci in commitems {diffitems.remove(ci)}

			return diffitems
		}

	}

	/**
		* DpchAppDo (full: DpchAppWzskNavOpDo)
		*/
	public class DpchAppDo: DpchAppWzsk {

		public static let SCRJREF: Int = 1
		public static let IXVDO: Int = 2
		public static let ALL: Int  = 3

		public init(
					_ scrJref: String = ""
					, _ ixVDo: Int = 0
					, mask: Set<Int> = []
				) {
			super.init(VecWzskVDpch.DPCHAPPWZSKNAVOPDO, scrJref)

			if mask.contains(DpchAppDo.ALL) {self.mask = [DpchAppDo.SCRJREF, DpchAppDo.IXVDO]}
			else {self.mask = mask}

			self.ixVDo = ixVDo
		}

		public var ixVDo: Int = 0

		public override func getSrefsMask() -> String {
			var ss: Array<String> = []

			if has(DpchAppDo.SCRJREF) {ss.append("scrJref")}
			if has(DpchAppDo.IXVDO) {ss.append("ixVDo")}

			return StrMod.vectorToString(ss)
		}

		public override func writeXML(
					_ doc: XMLDocument
					, _ sup: XMLElement?
				) {
			let el = XMLElement(name: "DpchAppWzskNavOpDo")

			if sup == nil {doc.setRootElement(el)}
			else {sup!.addChild(el)}

			el.addAttribute(XMLNode.attribute(withName: "xmlns", stringValue: "http://www.mpsitech.com") as! XMLNode)

			if has(DpchAppDo.SCRJREF) {el.addChild(XMLElement(name: "scrJref", stringValue: scrJref))}
			if has(DpchAppDo.IXVDO) {el.addChild(XMLElement(name: "srefIxVDo", stringValue: VecVDo.getSref(ixVDo)))}
		}

	}

	/**
		* DpchEngData (full: DpchEngWzskNavOpData)
		*/
	public class DpchEngData: DpchEngWzsk {

		public static let SCRJREF: Int = 1
		public static let STATAPP: Int = 2
		public static let STATSHR: Int = 3
		public static let TAG: Int = 4

		public init() {
			super.init(VecWzskVDpch.DPCHENGWZSKNAVOPDATA)
		}

		public var statapp: StatApp = StatApp()
		public var statshr: StatShr = StatShr()
		public var tag: Tag = Tag()

		public override func getSrefsMask() -> String {
			var ss: Array<String> = []

			if has(DpchEngData.SCRJREF) {ss.append("scrJref")}
			if has(DpchEngData.STATAPP) {ss.append("statapp")}
			if has(DpchEngData.STATSHR) {ss.append("statshr")}
			if has(DpchEngData.TAG) {ss.append("tag")}

			return StrMod.vectorToString(ss)
		}

		public override func readXML(
					_ doc: XMLDocument
					, _ _basexpath: String
					, _ addbasetag: Bool
				) {
			var valid: Bool
			var basexpath: String? = _basexpath

			clear()

			if addbasetag {basexpath = Xmlio.checkUclcXPaths(doc, _basexpath, "DpchEngWzskNavOpData")}
			valid = basexpath != nil

			if valid {valid = Xmlio.checkXPath(doc, basexpath!)}

			if valid {
				scrJref = Xmlio.extractStringUclc(doc, basexpath!, "scrJref", "", &mask, DpchEngData.SCRJREF)
				if statapp.readXML(doc, basexpath!, true) {add(DpchEngData.STATAPP)}
				if statshr.readXML(doc, basexpath!, true) {add(DpchEngData.STATSHR)}
				if tag.readXML(doc, basexpath!, true) {add(DpchEngData.TAG)}
			} else {
				scrJref = ""
				statapp = StatApp()
				statshr = StatShr()
				tag = Tag()
			}
		}

	}

}
